Edge scrolling: Configurable edge thickness

logiclrd opened the pull request:
This PR makes the thickness of the active scroll edge area for Edge Scrolling configurable.

The PR is formed of a series of logically-independent commits:

1. `CanvasModel` is updated to use a field instead of a local constant for the scroll edge thickness. Constants are added for minimum/maximum/default values.
2. UI elements are added to the Settings > Display section defined in `desktop_setting_page.dart` to represent the global default edge scroll edge thickness. A constant is added to represent the new setting in the underlying config. The UI has logic to automatically switch to ScrollEdge if the user interacts with the edge scroll edge thickness widgets, and to automatically submit changes to the text showing the edge thickness when it loses focus.
3. Preparatory changes in `remote_toolbar.dart`: The `_IconSubmenuButton` widget is reworked so that its `menuChildrenGetter` functor takes a pointer to the associated `_IconSubmenuButtonState`. A non-`@protected` `@override` of the `setState` method in `_IconSubmenuButtonState` is added, explicitly making the `setState` model for triggering UI rebuilds available to event handlers defined by `menuChildrenGetter`.
4. Preparatory changes in `remote_toolbar.dart`: The `_RectValueThumbShape` custom slider component shape is updated so that the units are configurable. They default to the '%' that was previously hardcoded, so as not to break any existing uses, but callers can now supply a custom `unit` value when constructing a `_RectValueThumbShape`.
5. UI elements are added to the Remote View toolbar defined in `remote_toolbar.dart` to represent the edge scroll edge thickness for the current session. The exposed `setState` method in the `state` parameter is used to trigger a rebuild when the user alters the current values in a way that keeps the menu pop-up open. The `ScrollEdge` menu item is updated to not close the menu pop-up on activation, allowing the user to further interact with the edge thickness controls. Also, if style other than `ScrollEdge` is selected and the user interacts with the edge scroll edge thickness slider, then the style is switched to `ScrollEdge` automatically. Underlying Rust methods for accessing the configuration in the "options" generic section of peer configuration are added, following the existing convention for other settings.
6. New strings are added to `src/lang/template.rs` and `res/lang.py` is rerun, propagating the new string blanks out to all the translations files. Values for English are provided.
7. An opportunistic fix to scroll-related code: Remove an unnecessary null assertion operator.
<img width="1186" height="893" alt="image" src="https://github.com/user-attachments/assets/76f2f516-ac74-4a7c-8cae-4d95aa9b6bed" />

`RustDesk_default.toml`:

```
[options]
scroll_style = 'scrolledge'
edge_scroll_edge_thickness = '108'
```

<img width="939" height="751" alt="image" src="https://github.com/user-attachments/assets/48f7330c-6f95-42b0-afa7-09f2847b67d9" />

```
MINGW64 ~/AppData/Roaming/RustDesk/config/peers
$ grep -B4 scroll 233165520.toml
    0,
    0,
]
view_style = 'original'
scroll_style = 'scrolledge'
--
[options]
zoom-cursor = ''
i444 = ''
custom-fps = '30'
edge_scroll_edge_thickness = '87'

MINGW64 ~/AppData/Roaming/RustDesk/config/peers
$
```

logiclrd committed:
Added constants kMinimumEdgeScrollEdgeThickness, kMaximumEdgeScrollEdgeThickness and kDefaultEdgeScrollEdgeThickness to consts.dart.
Added a field to CanvasModel in model.dart for the edge scroll edge thickness. Updated CanvasModel.edgeScrollMouse to use this field in place of local const edgeThickness.
Added method CanvasModel.updateEdgeScrollEdgeThickness to allow the value to be updated.
Updated updateScrollStyle and update in CanvasModel in model.dart to specify a Future<void> return type explicitly.

logiclrd committed:
Added constant kOptionEdgeScrollEdgeThickness to consts.dart to assign a name to a new config element.
Added edge scroll edge thickness to the scrollStyle widget generator in _DisplayState in desktop_setting_page.dart:
- Renamed the onChanged nested method to onScrollStyleChanged and extracted a separate method updateSavedScrollStyle for doing the configuration change.
- Updated the _Card with the radio button items to include a child item that presents a slider for updating the edge scroll edge thickness, with a label above it and a textbox with the raw value docked to the right. Added a FocusNode edgeScrollEdgeThicknessFocusNode to detect focus entering and exiting the text box.
- Added additional event handler methods automaticallySwitchToEdgeScroll, onEdgeScrollEdgeThicknessChanged and onEdgeScrollEdgeThicknessTextSubmitted, and added a listener to edgeScrollEdgeThicknessFocusNode.

logiclrd committed:
Updated the _IconSubmenuButton widget type in remote_toolbar.dart to include a state parameter of type _IconSubmenuButtonState in the menuChildrenGetter functor.
Updated all use sites correspondingly.
Updated the instantiation of the menu widgets in _iconSubmenuButtonState.build to pass in a pointer to itself when obtaining the widgets.
Added a non-@protected @override of _IconSubmenuButtonState.setState to allow menuChildrenGetter implementations to trigger rebuilds from their event handlers.

logiclrd committed:
Updated _RectValueThumbShape to have a configurable unit:
- Added field unit and updated the constructor to take it as an optional parameter defaulting to '%'.
- Updated _RectValueThumbShape.paint to use $unit instead of hardcoding a '%' character in the display value text.

logiclrd committed:
Added configuration accessor functions save_edge_scroll_edge_thickness and get_edge_scroll_edge_thickness to src/client.rs, src/flutter_ffi.rs and src/ui_session_interface.rs.
Added _IconSubmenuButtonState and ColorScheme arguments to the scrollStyle widget generator function in _DisplayMenuState. Updated the call site in the build method to supply the state given to the menuChildrenGetter method and the ColorScheme from the current Theme.
Updated scrollStyle to generate widgets for setting the edge scroll edge thickness.
Updated the prologue of the future returned by scrollStyle to obtain the edge scroll edge thickness setting value and supply it to the hasData stage.
Added a slider to the column of children returned by scrollStyle for setting the edge scroll edge thickness value.
Renamed the onChange handler in scrollStyle to onChangeScrollStyle and added a handler onChangeEdgeScrollEdgeThickness. Factored the configuration value save from onChangeScrollStyle to new method updateScrollStyle so that onChangeEdgeScrollEdgeThickness can switch the style to 'ScrollEdge' when the user interacts with the edge scroll edge thickness.
Updated the 'ScrollEdge' radio button in scrollStyle to not close the menu pop-up on activation, in order to allow the user to interact with the edge thickness slider. Updated onChangeScrollStyle to call state.setState to trigger a rebuild with the new value.

logiclrd committed:
Added new strings to src/lang/template.rs.
Reran res/lang.py to add the new strings to all languages.
Added English translations of the strings to src/lang/en.rs.

logiclrd committed:
Opportunistic: Remove unnecessary null assertion operator in updateScrollStyle in model.dart.

logiclrd committed:
PR feedback:
- Updated the scrollStyle widget generator in desktop_setting_page.dart to coalesce null edgeScrollEdgeThickness values to the default.
- Fixed copy/paste error in the inline documentation of get_edge_scroll_edge_thickness in src/client.rs.

rustdesk commented:
Could you move ScrollEdge under ScrollBar? I guess people prefer ScrollEdge than ScrollBar, but I do not want to change old users' behavior.

logiclrd committed:
PR feedback: Reorder scroll styles in desktop_setting_page.dart and remote_toolbar.dart.

21pages commented on src/client.rs:
```diff
@@ -1976,13 +1976,27 @@ impl LoginConfigHandler {
+    ///
+    /// * `value` - The edge thickness to be saved.
+    pub fn save_edge_scroll_edge_thickness(&mut self, value: i32) {
+        self.set_option("edge_scroll_edge_thickness".to_string(), value.to_string());
```
t is recommended to add a field in the `PeerConfig` of  crate `hbb_common` instead of using `options`. You can refer to `trackpad_speed`, 

https://github.com/rustdesk/rustdesk/blob/b75f4daa47bd8899549d15a22defe65bdfc28d6f/src/client.rs#L2408

https://github.com/rustdesk/hbb_common/blob/a4053b929b14059b1bd116900de8a103d9d838ae/src/config.rs#L331.

logiclrd replied:
Okay, that's cool. I was leaning toward doing that but previous advice in a related area had been to not add dedicated fields but instead use generic name/value functionality. But, I definitely prefer having a field for it because then it sits alongside `scroll_style` in the serialized form, much neater. :-) Will switch.

logiclrd replied:
Alright I have it all sorted locally. The `rustdesk` code, of course, can't work without the addition of the field to `hbb_common`, so I haven't yet pushed up the commit here. I'm waiting for the `hbb_common` change to go in before doing that. Let me know if I shouldn't wait and I'll go ahead and push this side of the changes too. 🙂

logiclrd replied:
Thinking it through, the only reason to not push the changes is that you can't build/test this branch without the `hbb_common` changes, but you can just go into the submodule repository and fetch the commit and check it out, so there's really no reason not to.

logiclrd committed:
Reworked save_edge_scroll_edge_thickness in LoginConfigHandler in src/client.rs to use the new PeerConfig::edge_scroll_edge_thickness property.
Updated get_edge_scroll_edge_thickness in the Session interface in src/ui_session_interface.rs to access edge_scroll_edge_thickness directly from the PeerConfig property instead of needing to call a get accessor.
Removed the unnecessary get_edge_scroll_edge_thickness accessor from LoginConfigHandler in src/client.rs.

21pages commented on flutter/lib/consts.dart:
```diff
@@ -79,6 +79,7 @@ const String kWindowEventOpenMonitorSession = "open_monitor_session";
 
 const String kOptionViewStyle = "view_style";
 const String kOptionScrollStyle = "scroll_style";
+const String kOptionEdgeScrollEdgeThickness = "edge_scroll_edge_thickness";
```
It is recommended to change it to `edge-scroll-edge-thickness`, as the use of underscores is for historical reasons.

logiclrd replied:
That feels weird directly adjacent to `view_style` and `scroll_style`, but I can of course make that change. 😛

logiclrd committed:
Updated the Rust code to account for the change to the edge scroll edge thickness value to a straight i32 instead of an Option<i32>:
- Updated save_edge_scroll_edge_thickness in src/client.rs to not wrap the supplied value in Option<>.
- Updated the return value of get_edge_scroll_edge_thickness in ui_session_interface.rs to match the underlying field type.
- Updated session_get_edge_scroll_edge_thickness in src/flutter_ffi.rs to wrap the edge scroll edge thickness value in Option<>, as the underlying value is now just a straight i32.
Updated src/flutter_ffi.rs to expose the UserDefaultConfig's number range mechanism:
- Added a locally-defined NumRange type, because rust-ffigen doesn't seem to be able to recognize an imported hbb_common::config::NumRange.
- Added method main_get_option_range to src/flutter_ffi.rs and method get_option_range to src/ui_interface.rs.
Updated kOptionEdgeScrollEdgeThickness in consts.dart to match keys::OPTION_EDGE_SCROLL_EDGE_THICKNESS in hbb_common, using hyphens instead of underscores.
Reworked desktop_setting_page.dart and remote_toolbar.dart to retrieve the NumRange for kOptionEdgeScrollEdgeThickness instead of directly referencing constants.
Deleted constants kMinimumEdgeScrollEdgeThickness, kMaximumEdgeScrollEdgeThickness and kDefaultEdgeScrollEdgeThickness from consts.dart.

21pages commented on flutter/lib/desktop/pages/desktop_setting_page.dart:
```diff
@@ -1678,29 +1678,111 @@ class _DisplayState extends State<_Display> {
+          groupValue: groupValue,
+          label: 'ScrollEdge',
+          onChanged: isOptFixed ? null : onScrollStyleChanged),
+      Container(
```
Display the slider only when 'ScrollEdge' is selected, follow the custom image quality setting.

logiclrd replied:
I like being able to select ScrollEdge by interacting with the slider, but I can switch it.

logiclrd commented on flutter/lib/desktop/widgets/remote_toolbar.dart:
```diff
@@ -1059,54 +1061,115 @@ class _DisplayMenuState extends State<_DisplayMenu> {
+      final edgeScrollEdgeThicknessRange = data['edgeScrollEdgeThicknessRange'] as NumRange;
+      updateScrollStyle(String value) async {
+        print("updateScrollStyle HAPPING");
+        print(StackTrace.current.toString());
```
Facepalm. I keep missing these. I do remove many of them while committing. Sigh. I should just add a step to search the source tree for `print` before I start committing. Sorry 🙂

logiclrd committed:
Moved the edge scroll edge thickness controls in desktop_setting_page.dart and in remote_toolbar.dart into Offstage widgets that hides them when the scroll style isn't kRemoteScrollStyleEdge.
Updated remote_toolbar.dart so that switching away from edge scrolling doesn't close the menu. Having the menu stay open when you switch to edge but then close if you immediately switch away feels wrong.
Removed an unnecessary await statement in remote_toolbar.dart.

logiclrd committed:
Removed debug output from remote_toolbar.dart. Oops

logiclrd committed:
Added method initializeEdgeScrollEdgeThickness to CanvasModel in model.dart to read the saved value from configuration.
Updated the first-image handlers in model.dart to call canvasModel.initializeEdgeScrollEdgeThickness.

logiclrd committed:
Reworked the handling of PeerConfig.custom_image_quality in client.rs to account for it being an i32 instead of a Vec<i32>.
Fixed the naming of NumRange.default/minimum/maximumValue in flutter_ffi.rs to default/minimum/maximum_value.
Updated the return value of session_get_custom_image_quality in flutter_ffi.rs to Option<i32>.
Updated the return value of get_custom_image_quality in ui_session_interface.rs to i32.
Removed unnecessary use of UserDefaultConfig in flutter_ffi.rs.
Updated the reference to UserDefaultConfig::get_num_range to be to an associated function rather than a member function, tracking change in hbb_common.
Added constast MAX_PUBLIC_SERVER_QUALITY for constraining quality when !allow_more in client.rs.
Updated customImageQualityWidget and customImageQualitySetting in setting_widgets.dart to receive NumRange values for the custom image quality range and the custom FPS range instead of using constants defined at the Dart level.
Updated customImageQualityDialog dialog.dart to receive NumRange values for the custom image quality range and the custom FPS range instead of using constants defined at the Dart level.
Updated toolbarImageQuality in toolbar.dart to retrieve the custom FPS range and the custom image quality range from the Rust layer and pass them into customImageQualityDialog. Added constants kOptionCustomImageQuality and kOptionCustomFps to consts.dart.
Updated DisplayState.build in desktop_setting_page.dart and DisplayPageState.build in settings_page.dart to retrieve the custom FPS range and the custom image quality range from the Rust layer and pass them into customImageQualitySetting.
Updated the handling of the customImageQuality setting in dialog.dart to receive an int instead of an Int32List.
Removed constants kMin/Default/MaxFps and kMin/Default/MaxQuality from consts.dart.

21pages commented on flutter/lib/common/widgets/toolbar.dart:
```diff
@@ -403,7 +406,7 @@ Future<List<TRadioMenu<String>>> toolbarImageQuality(
       onChanged: (value) {
         onChanged(value);
-        customImageQualityDialog(ffi.sessionId, id, ffi);
+        customImageQualityDialog(ffi.sessionId, id, ffi, customFpsRange, customQualityRange);
```
`customImageQualityDialog(SessionID sessionId, String id, FFI ffi, NumRange customQualityRange, NumRange customFpsRange)`

The order of the parameters needs to be adjusted.

rustdesk commented:
@21pages take over this and have a simple implementation.

logiclrd commented on flutter/lib/common/widgets/setting_widgets.dart:
```diff
@@ -6,24 +6,28 @@ import 'package:flutter_hbb/consts.dart';
+    required bool showMoreQuality,
+    required NumRange customQualityRange,
+    required NumRange customFpsRange}) {
+  if (initQuality < customQualityRange.minimumValue ||
```
Ah, right. When I first wrote part, it was in the mind that the custom quality range would be the restricted range, and so this logic was written with the thinking that it needs to override the standard range when "more" is enabled. But, when I was in the config.rs code, I realized that the standard range would have to be the extended range, otherwise it would never be able to read extended-range values back from the config file. I didn't remember that I had this logic here based on the wrong assumption.

logiclrd commented:
Thinking further about this, I think the real problem here is that I grew the scope of this PR. The actual edge thickness UI here is, I think, not more complex than #13445 (the other implementation is better because it centralizes the UI into a widget), but I felt an urge to map other things onto the updated paradigms, and doing so locally was good because it proved the concept, but those changes shouldn't be in this PR. This was an error on my part.
